### ISSUE TITLE
Support invoking callbacks with arbitrary timing

### ISSUE THREAD

**Issue by Schoonology** [testdouble/testdouble.js#106]
Since all callback invocations occur synchronously, any code that assumes synchronous execution will break when exposed to the Zalgo of the real world.

It would be nice if `td` could be configured to invoke callbacks with different timings, either synchronously or not. I don't have a solid suggestion for that API, but would be happy to think on it if y'all agree this would be useful.

**Comment by searls** [comment]
## Initial reaction

Agree with you that this is something worth considering. I'm, as usual, of two minds about this:

1. On one hand, td.js is designed to facilitate collaboration testing which isn't intended to provide rugged confidence for things like timing risks like this. I want to guard against adding features to td.js that would encourage testing patterns better left to a more integrated test.

2. On the other hand, it's entirely possible to test-drive these collaborators in such a way as to accidentally work synchronously, and since that's not how they'll ever be invoked in production, that's not good either.

3. On the third hand, I really like sync-by-default because I don't have to worry about silly `done` callbacking and so forth in my collaboration tests.

So where does that leave us? To be honest, I'm not really sure. 

## My first choice

My ideal, I think, would be a way to globally make all callbacks defer via some `setTimeout` or even `nextTick` as a way to just peek at the whole test suite every now and then that there aren't sync/async issues. However, that wouldn't really work, because:

1. most test frameworks have different semantics for sync tests from async tests, and each test would need to be updated for it
2. if there's a global option, some percentage of people will permanently toggle it on, despite the fact it's an unhealthy default (IMHO) for most uses and would be surprising to users visiting a project (as new tests might silently pass and the developer wouldn't know they need to code them as async)

## My second choice

The more practical option may be to add a `td.when` option, like you suggest, so that it's explicit and inline. I don't mind the [syntactic vinegar](http://david.heinemeierhansson.com/arc/2006_10.html) of making this hurt a bit, since I feel like it should be the exceptional case. The only problem is that most people will never see it or use it as a result, even though this is a risk that all callback-ey code faces.

I would propose:

``` js
td.when(schoon('hi'), {callbackTimeout: 5}).thenCallback(null, 'sup')
```

## My third choice

If this is a real problem and people are test-driving callback-APIs and bugs are introduced because td.js executes them synchronously, then I should give up my sync-test dream and just invoke the callbacks asynchronously to guard people against that risk by default. I really don't love this option but I'd do it if there's evidence that td.js is causing pain.

**Comment by rmg** [comment]
It's not so much the _time_ that matters, but whether it happens in the same call stack.

An API will/should always call the callback synchronously (`Array.prototype.map` and similar) or always call the callback asynchronously (anything involving IO, like a db query, _even if the result is cached_). Almost any Node app will have a mix of both types of APIs, so making a global toggle wouldn't be appropriate.

When I was reading through the docs and came across the note about `thenCallback` always being synchronous the first thing I did was look for `thenAsyncCallback` or `thenCallbackAsync`. I don't know about everyone, but that's what at least one person was expecting to see.

**Comment by searls** [comment]
Hey @rmg I think this is a valid line of reasoning, but I'm curious of some scenarios where deferring to a different call stack would matter to a user's code when invoked in a unit test context. Got a good, sellable example of where this won't work?

Separately, if we're talking options, I imagine I'd want:

```js
td.when(schoon('hi'), {defer: true}).thenCallback(null, 'sup')
```

**Comment by searls** [comment]
This just landed via #205

**Comment by rmg** [comment]
Thanks, @searls! I especially like the warnings you included in the README.md about how people probably don't need them for regular tests 👍
